### PR TITLE
Make regexes in types.cc static and remove unnecessary tolower transform

### PR DESCRIPTION
**Pull request by nuivall**
- makes all regexes static

If making regex compilation static
for uuid_type_impl and timeuuid_type_impl helps then it should
also help for timestamp_type and simple_date_type.

-  remove unnecessary tolower transform in simple_date_type_impl::from_sstring

Following function uses only decimal and '-' characters (see date_re). They are not
affected by tolower call in any way.

Aditionally std::strtoll supports "0x" prefixes but also accepts
upper case version "0X" so it's also not affected by tolower call.

get_simple_date_time only casts strings to integer types using
boost:lexical_cast so also not affected by tolower.

Finally, serialize only uses str to include it in an exception text
so tolower doesn't affect it in a positive way. It's even better
that input is displayed to the user as it was, not converted to lower
case.

### CODE REVIEW

File: types.cc [hunk 1]
```diff
@@ -352,14 +352,12 @@ static uint32_t serialize(const std::string& input, int64_t days) {
     return static_cast<uint32_t>(days);
 }
 uint32_t simple_date_type_impl::from_sstring(sstring_view s) {
-    std::string str;
-    str.resize(s.size());
-    std::transform(s.begin(), s.end(), str.begin(), ::tolower);
```
Comment (avikivity) [inline types.cc:357]: Why is this unnecessary? Changelog forgot to mention.

Comment (avikivity) [reply types.cc:357]: Well, I see the regex below doesn't mention any letters. But it would be nice for the changelog to say so.

Comment (nyh) [reply types.cc:357]: Even though the regex doesn't use letters, we end up calling other functions that parse parts of it - maybe those  that need to understand day name, month names, etc., do care about capitalization? I don't know, I am not familiar with any of these functions. @nuivall please investigate what these functions are actually doing, and explain in the commit message why the tolower (which has been in this code for years) is not needed.
Even better would be to add a unit test (in C++) to test this function. Or maybe we already have one?

**Review by nyh — CHANGES_REQUESTED** [review]
The first patch looks great, but Avi and I raised some doubts on the second patch (which may be correct, but it's not obvious that it is without further explanations or tests).

File: types.cc (hunk 1 above)
Comment (nuivall) [reply types.cc:357]: added explanation to the commit message

> Even better would be to add a unit test (in C++) to test this function.

Don't think it makes sense to add any here as I don't even know where behavior of this type is specified. If there are any violations in current behavior then existing test should catch it. There are some in test/cql-pytest/test_type_*.

The only place with docs I found is https://cassandra.apache.org/doc/latest/cassandra/cql/types.html#dates but according to it even this regex is not needed as format such as "-10-40-20" is not described there.

File: types.cc [hunk 2]
```diff
@@ -352,14 +352,12 @@ static uint32_t serialize(const std::string& input, int64_t days) {
...
     char* end;
     auto v = std::strtoll(s.begin(), &end, 10);
     if (end == s.begin() + s.size()) {
         return v;
     }
+    auto str = std::string(s); // FIXME: this copy probably can be avoided but needs some changes in regex_match
```
[the remark this answers is not in the document]
Comment (nuivall) [reply types.cc]: maybe it does work, although I am not sure how to code it, and even if I am not sure how it allocates memory for the results struct so simply leaving it as it was (still some small improvement as I moved this std::string allocation down, after strtoll fails to do the job)

we can't change std lib but there are other regex implementations, I don't know if it's accurate but some people on the internet are saying that regex library is not well maintained (e.g. https://www.reddit.com/r/cpp/comments/aqt7a0/status_of_string_view_and_regex/)

I added fixme as I think memory allocation/copy is not needed in this case, now I removed the part which suggests what to change (`needs some changes in regex_match`)

File: types.cc (hunk 1 above)
Comment (nyh) [reply types.cc:357]: I finally understand. This date type is just "yyyy-mm-dd", and is only numerals. There are no month names or anything like that. Ok.

**Review by nyh — APPROVED** [review]

File: types.cc (hunk 1 above)
Comment (nuivall) [reply types.cc:357]: ok, so we extend support for this from timestamp type to also date type.

Comment (nuivall) [reply types.cc:357]: or I misinterpret `For timestamps` wording in those docs...

Comment (nyh) [reply types.cc:357]: It's very unclear why they referred to timestamps :-(
When this happens to me, I usually write a test to see what's really going on... For example I wrote test/cql-pytest/test_type_timestamp.py for the "timestamp" type (don't confuse this with USING TIMESTAMP!), but never did this for the "date" type.

**Pull request merged**